### ISSUE TITLE
Feature importance return different value 

### ISSUE THREAD

**Issue by EdgarSanchez1796** [shap/shap#1090]
Hello, I am adjusting an xgboost model, after training I look at the importance of the variables for xgboost through gain, directly with the model.
  Then I get a plot summary to see the importance of the variables with shap, and the results are very different, variables that were in the middle of the table in xgboost appear in the first place in shap plot

**Comment by LEMTideman** [comment]
The "gain" measure of feature importance proposed by XGBoost is inconsistent and therefore unreliable (unlike Shapley Additive Explanations). Have a look at this Medium post by Scott Lundberg, which discusses your exact issue: https://towardsdatascience.com/interpretable-machine-learning-with-xgboost-9ec80d148d27 

Also interesting is the following paper "Consistent Individualized Feature Attribution for Tree Ensembles" by Scott M. Lundberg, Gabriel G. Erion, Su-In Lee. You can find it on Arxiv: https://arxiv.org/abs/1802.03888

Cheers, hope that was helpful!

**Comment by EdgarSanchez1796** [comment]
thanks for answering.
It's great, but I have a question, could I apply feature selection with the Shap value? I am doing this clearly with training, to avoid overfitting.

**Comment by LEMTideman** [comment]
SHAP is a post-hoc model agnostic interpretability method. Since it is a post-hoc method, it is applied to a machine learning model post-training, so I guess you could use it like a wrapper method in feature selection (not like a filter or embedded approach). SHAP was originally developed as a local interpretability method (i.e. for explaining the prediction made by the model on one specific data instance) but it can also be used as a global interpretability method. If you want to use SHAP for feature selection, I recommend using it as a global interpretability method: the function shap.summary_plot calculates the SHAP values per feature, and then averages their magnitudes over all data instances and presents you with a bar plot of feature importance (in descending order). You could then determine a threshold and only retain the most important features (i.e. maximum mean(|SHAP|)) as a way of performing selection. You would then need to train your model again using the new set of features.

**Comment by EdgarSanchez1796** [comment]
Exactly, I transformed the shap value  (based on the summary plot), cut by a certain value and then returned to training.

But I have an additional question, when you want to share the explanation of the model, with the influence graphs of each variable, do you recommend using training or test set to calculate the shap values?

**Thank you very much for your contribution to the community**


![Capture](https://user-images.githubusercontent.com/24661488/76327278-9326db80-62c8-11ea-86d0-558f7222df3b.PNG)

**Comment by LEMTideman** [comment]
Hello!
You could use SHAP to measure the predictive importance of your features either on your training set or on your testing set. Applying SHAP to your training set will help you inspect the ML model, and better understand the model's decision-making process ("opening up the black box"). However, it is only by applying SHAP to your testing set that you will be able to figure out how the features impact the model's generalization performance, so I recommend applying SHAP to the testing set.

**Comment by EdgarSanchez1796** [comment]
Hi, i have an other similar question, it's possible obtain the same probability that it's estimated with de xgboost model?
i am using the expected value and logit, but it's very different
Thanks!

**Comment by alexcoca** [comment]
Are you using the `TreeExplainer`  or the `KernelExplainer` @EdgarSanchez1796 ? Is it a classification or a regression problem?
